core/src/main/java/org/jboss/logmanager/MDC.java

```diff
@@ -19,16 +19,32 @@
 
 package org.jboss.logmanager;
 
+import java.util.Iterator;
 import java.util.Map;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
 
 /**
  * Mapped diagnostic context.  This is a thread-local map used to hold loggable information.
  */
 public final class MDC {
+    private static final MDCProvider mdcProvider = getMDCProvider();
 
     private MDC() {}
 
-    private static final Holder mdc = new Holder();
+    static MDCProvider getMDCProvider() {
+        final ServiceLoader<MDCProvider> configLoader = ServiceLoader.load(MDCProvider.class, MDC.class.getClassLoader());
+        final Iterator<MDCProvider> iterator = configLoader.iterator();
+        for (;;) try {
+            if (! iterator.hasNext()) {
+                return new ThreadLocalMDC();
+            }
+            return iterator.next();
+        } catch (ServiceConfigurationError | RuntimeException e) {
+            System.err.print("Warning: failed to load MDC Provider: ");
+            e.printStackTrace(System.err);
+        }
+    }
```

Comment:
Should we do this in a privileged block? Or use a try/catch and return the default maybe.

Reply:
Yeah, good point: these changes were all designed for Quarkus which doesn't support security managers. I'll revise it tomorrow to add `doPrivileged` when the SM is installed.

There already is a fallback to the default in the case that no services are loadable.

Reply:
Okay cool. If it's wrapped in a `doPrivileged()` the fallback is fine.

Reply:
Fixed.